TXT GROUP WEEK 2: LISTING, SPLITING, AND SLICING; or, the Things We Do in Brackets

Last session, we went over how to define a VARIABLE, what different types of data Python accepts, and how to set a conditional in the form of an IF STATEMENT.

To review these quickly, we’re going to define the VARIABLE topic as “Letters”, result as “?”, and book_index as an especially long string.

In [6]:
topic = "Letters"
results = "?"
book_index = "Chapter 1: Conversation Chapter 2: Dress Chapter 3: Traveling Chapter 4: How to Behave at a Hotel Chapter 5: Letters of Business Chapter 6: Letters of the Heart Chapter 7: Evening Parties as Hostess"

We want to see if our topic appears within book_index to find which chapter we might want to reference.

In [4]:
if topic in book_index:
	results = "Yeah, there's a result. Or results? I really don't know"
else: 
	results = "Nah, I don't think there are any results."

print(results)

Yeah, there's a result. Or results? I really don't know


As the results show us above, this current if statement wouldn't be able to distinguish whether the word "Letters" appears one or a million times in book_index; it is satisfied after the first hit. By using LISTS and SLICES, we're going to write code that will give us significantly more informative results.

SLICING

If we're going to build code that will give us more robust information, we will first need to know what the infomation we want is. One thing we might want to know is which exact chapters match our topic. By scanning the “book_index” string, our eyes might be able to quickly tell which chapter we want. Typographically, the colon after each chapter number states what each chapter covers. There are seven chapters in book_index and two of those chapters deal with letters.

In [43]:
print(book_index)

Chapter 1: Conversation Chapter 2: Dress Chapter 3: Traveling Chapter 4: How to Behave at a Hotel Chapter 5: Letters of Business Chapter 6: Letters of the Heart Chapter 7: Evening Parties as Hostess


We may see seven chapters, but if we run a len() command on book_index, Python outputs a very different number.

In [8]:
len(book_index)

198

If we want to know what those 198 components are we can SLICE through book_index. Much like a piece of pie, a SLICE is a portion of a whole. We can SLICE however large a piece that we'd like in whatever direction we prefer. SLICING is marked by brackets [ ].

In [14]:
book_index[0:10]
#try substituting in different numbers. What happens when you leave certain coordinates blank?#

'Chapter 1:'

In [15]:
book_index[-10:]

'as Hostess'

So, if we want to know what the first item in book_index is, we would type book_index[0]. if we wanted to know the last part of book_index, we would write book_index[-1] If we only wanted to see a section of book index, we would SLICE as follows: book_index [3:8]

This may not seem that important given that book_index is pretty readable. But let’s imagine instead that “book_index” is actually a collection of indices from eight hundred self-help books. Having all of this data as a single, readable string is untenable. Even if we used command/control F, we wouldn’t intuitively know which book we were looking at either.

SPLITING

Thankfully, Python has a built in function for splitting a data type by common patterns. The .split() commands allows users to take a string and split it based on a feature. Which word or character would we want to SPLIT book_index by if we wanted the results len() to be 7? Or, which word occurs seven times times in valuable places?

In [26]:
my_birthday = "11/18/1991"
my_birthday_n = my_birthday.split("/")
b_month = my_birthday_n[0]
b_date = my_birthday_n[1]
b_year = my_birthday_n[2]

print(b_month)

11


In [46]:
#finish the following code#
book_index_n = book_index.split("Chapter")
len(book_index_n)

8

That should have worked! Why does the result say that len = 8? We have a blank entry. There are a few ways that we can get rid of this entry, but I would like to focus on one method that uses slicing. If the first entry is blank, how could we use slicing to only have relevant entries.

In [47]:
#finish the following code#
book_index_n = book_index_n[1:]

type(book_index_n)

list

You may notice that when we slice through step_one, some entires contain far more than a single character. Using the type() function, we find out that type(step_one) is a list, where as topic and book_index are strings.

LISTS

A LIST is a scalable collection of similar types of data. We define a LIST through using brackets. While we can have lists of integers and lists of strings, the two do not mix. In the example below, some strings are converted into integers for mathematical operations and then converted back again.

In [39]:
my_favorite_things = ["funky patterned shirts", "homemade bread", "anything woolen", "bathbombs", "The Cure or The Cramps depending on mood", "pancakes", "novels with dancing and scandal"]
lucky_numbers = [4, 21, 72]

Todays_month = 9

if Todays_month == int(b_month):
    print("Malcolm likes " + my_favorite_things[2])
else:
    print("Only " + str(abs(int(b_month) - Todays_month)) + " more months until Malcolm's birthday ;)")


Only 2 more months until Malcolm's birthday ;)


One important quality to lists is that they’re mutable. We can add new entries to a list and even change existing entries.

We’re going to test this mutability by adding the word “Chapter” back to each entry. Test the following two pieces of code. Why does one work and the other does not?

In [48]:
for entry in book_index_n:
	entry = "Chapter" + entry

print(book_index_n)


[' 1: Conversation ', ' 2: Dress ', ' 3: Traveling ', ' 4: How to Behave at a Hotel ', ' 5: Letters of Business ', ' 6: Letters of the Heart ', ' 7: Evening Parties as Hostess']


In [49]:
for number in range(len(book_index_n)):
	book_index_n[number] = "Chapter" + book_index_n[number]
    
print(book_index_n)

['Chapter 1: Conversation ', 'Chapter 2: Dress ', 'Chapter 3: Traveling ', 'Chapter 4: How to Behave at a Hotel ', 'Chapter 5: Letters of Business ', 'Chapter 6: Letters of the Heart ', 'Chapter 7: Evening Parties as Hostess']


If we wanted to delimit these results further, we could use the split command once again.

In [52]:
book_index_BEST = []

for item in book_index_n:
    ch = item.split(":")[0]
    des = item.split(":")[1]
    book_index_BEST.append([ch, des])
    
print(book_index_BEST)

[['Chapter 1', ' Conversation '], ['Chapter 2', ' Dress '], ['Chapter 3', ' Traveling '], ['Chapter 4', ' How to Behave at a Hotel '], ['Chapter 5', ' Letters of Business '], ['Chapter 6', ' Letters of the Heart '], ['Chapter 7', ' Evening Parties as Hostess']]


Now that we have our list up and running, we can be a bit more ambitious with our research. Instead of searching only for a single topic, let’s look at multiple topics. Our first step is going to be changing our variable from topic to topics. Now it’s official. Next, we’ll want to add our new topics, using the append function of Python.


In [54]:
topics = topic
topics.append("Parties", "Conduct", "Etiquette", "Dainty H'ors D'ouevres")
topics

AttributeError: 'str' object has no attribute 'append'

Why didn’t that work? For one, pluralizing topic to topics didn’t change it from a single string to a list. Making a list in Python requires brackets.

In [55]:
topics = [topic]
topics.append("Parties", "Conduct", "Etiquette", "Dainty H'ors D'ouevres")
topics

TypeError: append() takes exactly one argument (4 given)

So let’s try this again, now that we’re positive that topics is a list. Why didn’t that work? The append command adds only a single item to a list at a time. We got greedy and tried to add too many at once. Working either alone or with a partner, come up with a way to add these new_topics to the list topics one by one.

In [56]:
for new_topic in ["Parties", "Conduct", "Etiquette", "Dainty H'ors D'ouevres"]:
	topics.append(new_topic)
topics


['Letters', 'Parties', 'Conduct', 'Etiquette', "Dainty H'ors D'ouevres"]

Now let's put it all together!

In [60]:
for topic in topics:
	for word in book_index_BEST:
		if topic in word[1]:
			print(word[0])

Chapter 5
Chapter 6
Chapter 7
